### PR TITLE
[DPE-4974] Use ruff + dpw v17.0.1 + poetry + enable renovate + set up allure

### PR DESCRIPTION
**Pull request by shayancanonical**
## Issue
This repository is way behind modern CI in the following ways:
1. It does not use poetry for dependency management
2. Does not use ruff for linting and formatting
3. Does not use data-platform-workflows v17.0.1
4. Does not use renovate for dependency upgrade
5. We dont have allure for this repository

## Solution
1. Use poetry
2. Use ruff
3. Use dpw 17.0.1
4. Use renovate
5. Enable allure

## Side effects
1. There will be allure reports for nightly scheduled tests 
2. The update-bundles workflow will correctly account for amd64 vs arm64 revisions when updating bundles

### CODE REVIEW

**Review by paulomach — APPROVED** [review]
much better

File: tox.ini [hunk 1]
```diff
@@ -15,48 +15,43 @@ set_env =
...
-    flake8-copyright
-    flake8-builtins
-    pyproject-flake8
-    pep8-naming
-    isort
-    codespell
+allowlist_externals =
+    {[testenv]allowlist_externals}
+commands_pre =
+    poetry install --only lint,format --no-root
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tox.ini]: addressed in 09d0fdfc0a7ba6df72168667af84425057257478

File: tox.ini [hunk 2]
```diff
@@ -15,48 +15,43 @@ set_env =
...
+    # Workaround for https://github.com/python-poetry/poetry/issues/6958
+    POETRY_INSTALLER_PARALLEL = false
+pass_env =
+    CI
+    GITHUB_OUTPUT
+    SECRETS_FROM_GITHUB
+allowlist_externals =
+    poetry
+commands_pre =
+    poetry install --only integration --no-root
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tox.ini:48]: addressed in 09d0fdfc0a7ba6df72168667af84425057257478

File: tox.ini [hunk 3]
```diff
@@ -15,48 +15,43 @@ set_env =
...
-    flake8
-    flake8-docstrings
-    flake8-copyright
-    flake8-builtins
-    pyproject-flake8
-    pep8-naming
-    isort
-    codespell
+allowlist_externals =
+    {[testenv]allowlist_externals}
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tox.ini]: addressed in 09d0fdfc0a7ba6df72168667af84425057257478

File: tox.ini [hunk 4]
```diff
@@ -15,48 +15,43 @@ set_env =
...
+allowlist_externals =
+    poetry
 
 [testenv:format]
 description = Apply coding style standards to code
-deps =
-    black
-    isort
+commands_pre =
+    poetry install --only format --no-root
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tox.ini]: addressed in 09d0fdfc0a7ba6df72168667af84425057257478

File: tox.ini
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tox.ini:1]: addressed in 09d0fdfc0a7ba6df72168667af84425057257478

File: tox.ini [hunk 5]
```diff
@@ -15,48 +15,43 @@ set_env =
     PYTHONPATH = {tox_root}/lib
     PYTHONBREAKPOINT=ipdb.set_trace
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tox.ini]: addressed in 09d0fdfc0a7ba6df72168667af84425057257478

File: pyproject.toml [hunk 6]
```diff
@@ -15,27 +49,31 @@ asyncio_mode = "auto"
 markers = ["unstable"]
 
 # Formatting tools configuration
-[tool.black]
+[tool.ruff]
+# preview and explicit preview are enabled for CPY001
+preview = true
+target-version = "py38"
+src = ["src", "."]
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: addressed in 09d0fdfc0a7ba6df72168667af84425057257478

File: pyproject.toml [hunk 7]
```diff
@@ -1,6 +1,40 @@
...
+[tool.poetry.group.integration.dependencies]
+pytest = "^8.3.2"
+pytest-github-secrets = {git = "https://github.com/canonical/data-platform-workflows", tag = "v17.0.1", subdirectory = "python/pytest_plugins/github_secrets"}
+pytest-operator = "^0.35.0"
+pytest-operator-cache = {git = "https://github.com/canonical/data-platform-workflows", tag = "v17.0.1", subdirectory = "python/pytest_plugins/pytest_operator_cache"}
+pytest-operator-groups = {git = "https://github.com/canonical/data-platform-workflows", tag = "v17.0.1", subdirectory = "python/pytest_plugins/pytest_operator_groups"}
+allure-pytest-collection-report = {git = "https://github.com/canonical/data-platform-workflows", tag = "v17.0.1", subdirectory = "python/pytest_plugins/allure_pytest_collection_report"}
+# renovate caret doesn't work: https://github.com/renovatebot/renovate/issues/26940
+juju = "<=3.5.2.0"
+mysql-connector-python = "8.0.33"
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: addressed in 09d0fdfc0a7ba6df72168667af84425057257478

File: pyproject.toml [hunk 8]
```diff
@@ -1,6 +1,40 @@
...
+
+[tool.poetry.group.integration.dependencies]
+pytest = "^8.3.2"
+pytest-github-secrets = {git = "https://github.com/canonical/data-platform-workflows", tag = "v17.0.1", subdirectory = "python/pytest_plugins/github_secrets"}
+pytest-operator = "^0.35.0"
+pytest-operator-cache = {git = "https://github.com/canonical/data-platform-workflows", tag = "v17.0.1", subdirectory = "python/pytest_plugins/pytest_operator_cache"}
+pytest-operator-groups = {git = "https://github.com/canonical/data-platform-workflows", tag = "v17.0.1", subdirectory = "python/pytest_plugins/pytest_operator_groups"}
+allure-pytest-collection-report = {git = "https://github.com/canonical/data-platform-workflows", tag = "v17.0.1", subdirectory = "python/pytest_plugins/allure_pytest_collection_report"}
+# renovate caret doesn't work: https://github.com/renovatebot/renovate/issues/26940
+juju = "<=3.5.2.0"
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: good catch, addressed in 09d0fdfc0a7ba6df72168667af84425057257478

File: pyproject.toml [hunk 9]
```diff
@@ -1,6 +1,40 @@
...
+
+[tool.poetry.group.lint]
+optional = true
+
+[tool.poetry.group.lint.dependencies]
+ruff = "^0.5.5"
+codespell = "2.3.0"
+
+[tool.poetry.group.integration]
+optional = true
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: addressed in 09d0fdfc0a7ba6df72168667af84425057257478

File: pyproject.toml [hunk 10]
```diff
@@ -1,6 +1,40 @@
 # Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
+[tool.poetry]
+package-mode = false
+
+[tool.poetry.dependencies]
+python = "^3.8.10"
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: addressed in 09d0fdfc0a7ba6df72168667af84425057257478

File: .github/workflows/ci.yaml [hunk 11]
```diff
@@ -16,33 +16,28 @@ on:
...
-        uses: charmed-kubernetes/actions-operator@main
-        with:
-          provider: microk8s
-          channel: 1.30-strict/stable
-          bootstrap-options: "--agent-version 3.5.2"
-      - name: Run integration bundle test
-        run: tox -e integration
+    uses: canonical/data-platform-workflows/.github/workflows/integration_test_charm.yaml@v17.0.1
+    with:
+      artifact-prefix: packed-charm-
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply .github/workflows/ci.yaml]: addressed in 09d0fdfc0a7ba6df72168667af84425057257478

File: .github/workflows/ci.yaml [hunk 12]
```diff
@@ -16,33 +16,28 @@ on:
...
       fail-fast: false
-    name: Integration tests (K8s)
+      matrix:
+        juju:
+          - agent: 3.5.3  # renovate: juju-agent-pin-minor
+            juju-snap-channel: 3.5/stable
+            allure_on_amd64: true
+        architecture:
+          - amd64
+    name: Integration test bundle | ${{ matrix.juju.agent }}
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply .github/workflows/ci.yaml]: addressed in 09d0fdfc0a7ba6df72168667af84425057257478

File: .github/workflows/ci.yaml [hunk 13]
```diff
@@ -16,33 +16,28 @@ on:
...
     strategy:
       fail-fast: false
-    name: Integration tests (K8s)
+      matrix:
+        juju:
+          - agent: 3.5.3  # renovate: juju-agent-pin-minor
+            juju-snap-channel: 3.5/stable
+            allure_on_amd64: true
+        architecture:
+          - amd64
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply .github/workflows/ci.yaml]: removed in 09d0fdfc0a7ba6df72168667af84425057257478

File: .github/renovate.json5 [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+  "reviewers": ["paulomach", "carlcsaposs-canonical", "shayancanonical", "taurus-forever"],
+  "packageRules": [
+    // Later rules override earlier rules
+    {
+      "matchPackageNames": ["pydantic"],
+      "allowedVersions": "<2.0.0"
+    }
+  ],
+  "regexManagers": [
+  ]
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply .github/renovate.json5]: removed in 09d0fdfc0a7ba6df72168667af84425057257478

File: .github/renovate.json5 [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
+  "extends": ["github>canonical/data-platform//renovate_presets/charm.json5"],
+  "reviewers": ["paulomach", "carlcsaposs-canonical", "shayancanonical", "taurus-forever"],
+  "packageRules": [
+    // Later rules override earlier rules
+    {
+      "matchPackageNames": ["pydantic"],
+      "allowedVersions": "<2.0.0"
+    }
+  ],
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply .github/renovate.json5]: no pydantic in repo. removed in 09d0fdfc0a7ba6df72168667af84425057257478

File: .github/renovate.json5 [hunk 16]
```diff
@@ -0,0 +1,14 @@
+{
+  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
+  "extends": ["github>canonical/data-platform//renovate_presets/charm.json5"],
+  "reviewers": ["paulomach", "carlcsaposs-canonical", "shayancanonical", "taurus-forever"],
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply .github/renovate.json5]: changed in 09d0fdfc0a7ba6df72168667af84425057257478

**Review by taurus-forever — APPROVED** [review]
Thank you!

File: .github/workflows/ci.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (shayancanonical) [reply .github/workflows/ci.yaml]: didnt understand what you meant by no need for matrix until i stared a bit longer. should be addressed in 29290319e7a89181aba508bf25ed67e53bfce1f2

File: tox.ini (hunk 2 above)
[the remark this answers is not in the document]
Comment (shayancanonical) [reply tox.ini]: removed in 29290319e7a89181aba508bf25ed67e53bfce1f2

File: .github/workflows/ci.yaml [hunk 17]
```diff
@@ -16,33 +16,21 @@ on:
...
-      - name: Install tox
-        # TODO: Consider replacing with custom image on self-hosted runner OR pinning version
-        run: python3 -m pip install tox
-      - name: Run linters
-        run: tox run -e lint
+    uses: canonical/data-platform-workflows/.github/workflows/lint.yaml@v17.0.1
 
   integration-test:
     strategy:
       fail-fast: false
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply .github/workflows/ci.yaml]: addressed in 1ce5af1cad03eca4f27311bc9bb1fa7720c8feab

**Pull request merged**